gymnasium/utils/serialize_spec_stack.py

```diff
@@ -0,0 +1,139 @@
+"""Functions that use/modify specification stacks."""
+import dataclasses
+import inspect
+import json
+import re
+from typing import Union
+
+from gymnasium import error
+from gymnasium.envs.registration import EnvSpec
+from gymnasium.wrapperspec import WrapperSpec
+
+
+def serialise_spec_stack(stack: "tuple[Union[WrapperSpec, EnvSpec]]") -> str:
+    """Serialises the specification stack into a JSON string.
+
+    Args:
+        stack: Tuple of environment and wrapper specifications, known as the specification stack. Generated via `env.spec_stack`.
+
+    Returns:
+        A JSON string representing the specification stack.
+    """
+    num_layers = len(stack)
+    stack_json = {}
+    for i, spec in enumerate(stack):
+        spec = _serialise_callable(spec)
+        if i == num_layers - 1:
+            if isinstance(spec, WrapperSpec) or not isinstance(
+                spec, EnvSpec
+            ):  # NB: WrapperSpec is a subclass of EnvSpec
+                raise error.Error(
+                    f"Expected spec to be an EnvSpec, but got {type(spec)} instead."
+                )
+            layer = "raw_env"
+        else:
+            if not isinstance(spec, WrapperSpec):
+                raise error.Error(
+                    f"Expected spec to be a WrapperSpec, but got {type(spec)} instead."
+                )
+            layer = f"wrapper_{num_layers - i - 2}"
+        spec_json = json.dumps(dataclasses.asdict(spec))
+        stack_json[layer] = spec_json
+    return stack_json
+
+
+def _serialise_callable(spec: Union[WrapperSpec, EnvSpec]) -> dict:
+    """Converts a spec's lambda functions to a string representation.
+
+    Args:
+        spec: An environment or wrapper specification.
+
+    Returns: The specification with lambda functions converted to strings.
+
+    """
+    for k, v in spec.kwargs.items():
+        if callable(v):
+            try:
+                str_repr = (
+                    str(inspect.getsourcelines(v)[0]).strip("['\\n']").split(" = ")[1]
+                )  # https://stackoverflow.com/a/30984012
+            except OSError:
+                raise error.Error(
+                    "The attempted seialisation of a callable failed. This is likely due to env.spec_stack being called twice, which for technical reasons doesn't work. Please modify your code to only call env.spec_stack once (you can save that to a variable and repeatedly use that)."
```

Comment:
u(h)*100

Is there like a deep technical reason why this happens? I'm guessing it's related to the mildly cursed parsing of the lambda function code?

Reply:
This also fails in an interactive context which is a huge use-case of python

Reply:
This also also fails with an input like
```python
env = gym.experimental.wrappers.LambdaActionV0(env, lambda x: 1-x, None)
```
where the lambda function is parsed as 
```python
lambda x: 1-x, None
```

I didn't actually check deserialization, but I can't see this working properly

Reply:
Also it completely fails (with an incomprehensible error) when the lambda function is defined with the usual
```python
def switch(action):
    return 1 - action
```

AFAIK (or I definitely hope so) the lambda wrappers aren't limited to the actual python lambda statements, which are... bad. 
At least my understanding and intention is that the "lambda" refers to the general idea of pure functions (going back to **lambda** calculus), and overfitting to the specific way of defining pure functions calls for huge trouble.